diff --git a/src/globals.h b/src/globals.h
index d5320d7..968ba33 100644
--- a/src/globals.h
+++ b/src/globals.h
@@ -1657,6 +1657,11 @@ EXTERN int *eval_lavars_used INIT(= NULL);
 EXTERN int ctrl_break_was_pressed INIT(= FALSE);
 #endif
 
+#ifdef FEAT_SPELL
+EXTERN char e_illegal_character_in_word[]
+	INIT(= N_("E1280: Illegal character in word"));
+#endif
+
 /*
  * Optional Farsi support.  Include it here, so EXTERN and INIT are defined.
  */
diff --git a/src/mbyte.c b/src/mbyte.c
index 6d21f11..a7531f1 100644
--- a/src/mbyte.c
+++ b/src/mbyte.c
@@ -4034,7 +4034,7 @@ theend:
     convert_setup(&vimconv, NULL, NULL);
 }
 
-#if defined(FEAT_GUI_GTK) || defined(PROTO)
+#if defined(FEAT_GUI_GTK) || defined(FEAT_SPELL) || defined(PROTO)
 /*
  * Return TRUE if string "s" is a valid utf-8 string.
  * When "end" is NULL stop at the first NUL.
diff --git a/src/spellfile.c b/src/spellfile.c
index 496e07f..92997ef 100644
--- a/src/spellfile.c
+++ b/src/spellfile.c
@@ -4441,6 +4441,10 @@ store_word(
     int		res = OK;
     char_u	*p;
 
+    // Avoid adding illegal bytes to the word tree.
+    if (enc_utf8 && !utf_valid_string(word, NULL))
+	return FAIL;
+
     (void)spell_casefold(word, len, foldword, MAXWLEN);
     for (p = pfxlist; res == OK; ++p)
     {
@@ -6251,6 +6255,12 @@ spell_add_word(
     int		i;
     char_u	*spf;
 
+    if (enc_utf8 && !utf_valid_string(word, NULL))
+    {
+	EMSG(_(e_illegal_character_in_word));
+	return;
+    }
+
     if (idx == 0)	    /* use internal wordlist */
     {
 	if (int_wordlist == NULL)
