diff -up vim91/runtime/doc/options.txt.modeline-tests vim91/runtime/doc/options.txt
--- vim91/runtime/doc/options.txt.modeline-tests	2024-02-09 06:33:54.000000000 +0100
+++ vim91/runtime/doc/options.txt	2026-04-08 15:14:10.424275654 +0200
@@ -2070,6 +2070,8 @@ A jump table for the options with a shor
 	based expansion (e.g., dictionary |i_CTRL-X_CTRL-K|, included patterns
 	|i_CTRL-X_CTRL-I|, tags |i_CTRL-X_CTRL-]| and normal expansions).
 
+	This option cannot be set in a modeline when 'modelineexpr' is off.
+
 						*'completefunc'* *'cfu'*
 'completefunc' 'cfu'	string	(default: empty)
 			local to buffer
@@ -4177,6 +4179,7 @@ A jump table for the options with a shor
 	You can include a line break.  Simplest method is to use |:let|: >
 		:let &guitabtooltip = "line one\nline two"
 <
+	This option cannot be set in a modeline when 'modelineexpr' is off.
 
 						*'helpfile'* *'hf'*
 'helpfile' 'hf'		string	(default (MS-Windows) "$VIMRUNTIME\doc\help.txt"
@@ -6238,6 +6241,8 @@ A jump table for the options with a shor
 	The format of the header produced in |:hardcopy| output.
 	See |pheader-option|.
 
+	This option cannot be set in a modeline when 'modelineexpr' is off.
+
 						*'printmbcharset'* *'pmbcs'*
 'printmbcharset' 'pmbcs'  string (default "")
 			global
diff -up vim91/src/testdir/test_modeline.vim.modeline-tests vim91/src/testdir/test_modeline.vim
--- vim91/src/testdir/test_modeline.vim.modeline-tests	2026-04-08 15:12:08.284334428 +0200
+++ vim91/src/testdir/test_modeline.vim	2026-04-08 15:14:50.248503197 +0200
@@ -262,13 +262,16 @@ endfunc
 
 func Test_modeline_fails_modelineexpr()
   call s:modeline_fails('balloonexpr', 'balloonexpr=Something()', 'E992:')
+  call s:modeline_fails('complete', "complete=FSomething", 'E992:')
   call s:modeline_fails('foldexpr', 'foldexpr=Something()', 'E992:')
   call s:modeline_fails('foldtext', 'foldtext=Something()', 'E992:')
   call s:modeline_fails('formatexpr', 'formatexpr=Something()', 'E992:')
   call s:modeline_fails('guitablabel', 'guitablabel=Something()', 'E992:')
+  call s:modeline_fails('guitabtooltip', 'guitabtooltip=Something()', 'E992:')
   call s:modeline_fails('iconstring', 'iconstring=Something()', 'E992:')
   call s:modeline_fails('includeexpr', 'includeexpr=Something()', 'E992:')
   call s:modeline_fails('indentexpr', 'indentexpr=Something()', 'E992:')
+  call s:modeline_fails('printheader', 'printheader=Something()', 'E992:')
   call s:modeline_fails('rulerformat', 'rulerformat=Something()', 'E992:')
   call s:modeline_fails('statusline', 'statusline=Something()', 'E992:')
   call s:modeline_fails('tabline', 'tabline=Something()', 'E992:')
@@ -362,29 +365,4 @@ func Test_modeline_disable()
   call assert_equal(2, &sw)
 endfunc
 
-func Test_modeline_forbidden()
-  let tempfile = tempname()
-  let lines =<< trim END
-    some test text for completion
-    vim: set complete=F{->system('touch_should_not_run')} :
-  END
-  call writefile(lines, tempfile, 'D')
-  call assert_fails($'new {tempfile}', 'E992:')
-  bw!
-  let lines =<< trim END
-    some text
-    vim: set guitabtooltip=%{%mapset()%}:
-  END
-  call writefile(lines, tempfile)
-  call assert_fails($'new {tempfile}', 'E992:')
-  bw!
-  let lines =<< trim END
-    some text
-    vim: set printheader=%{mapset('n',0,{})%)%}:
-  END
-  call writefile(lines, tempfile, 'D')
-  call assert_fails($'new {tempfile}', 'E992:')
-  bw!
-endfunc
-
 " vim: shiftwidth=2 sts=2 expandtab
